plugin/pkg/shootstate/admission.go

```diff
@@ -0,0 +1,177 @@
+func New() (*ValidateShootStateDeletion, error) {
+	return &ValidateShootStateDeletion{
+		Handler: admission.NewHandler(admission.Delete),
+	}, nil
+}
+
+// AssignReadyFunc assigns the ready function to the admission handler.
+func (d *ValidateShootStateDeletion) AssignReadyFunc(f admission.ReadyFunc) {
+	d.readyFunc = f
+	d.SetReadyFunc(f)
+}
+
+// SetExternalCoreInformerFactory sets the external garden core informer factory.
+func (d *ValidateShootStateDeletion) SetExternalCoreInformerFactory(f externalcoreinformers.SharedInformerFactory) {
+	shootStateInformer := f.Core().V1alpha1().ShootStates()
+	d.shootStateLister = shootStateInformer.Lister()
+
+	readyFuncs = append(readyFuncs, shootStateInformer.Informer().HasSynced)
+}
+
+// SetInternalCoreClientset sets the garden core clientset.
+func (d *ValidateShootStateDeletion) SetInternalCoreClientset(c coreclientset.Interface) {
+	d.coreClient = c
+}
+
+func (d *ValidateShootStateDeletion) waitUntilReady(attrs admission.Attributes) error {
+	// Wait until the caches have been synced
+	if d.readyFunc == nil {
+		d.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+
+	if !d.WaitForReady() {
+		return admission.NewForbidden(attrs, errors.New("not yet ready to handle request"))
+	}
+
+	return nil
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (d *ValidateShootStateDeletion) ValidateInitialization() error {
+	if d.shootStateLister == nil {
+		return errors.New("missing ShootState lister")
+	}
+
+	if d.coreClient == nil {
+		return errors.New("missing garden core client")
+	}
+
+	return nil
+}
+
+var _ admission.ValidationInterface = &ValidateShootStateDeletion{}
+
+// Validate makes admissions decisions based on deletion confirmation annotation.
+func (d *ValidateShootStateDeletion) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	if a.GetKind().GroupKind() != core.Kind("ShootState") {
+		return nil
+	}
+
+	if err := d.waitUntilReady(a); err != nil {
+		return fmt.Errorf("err while waiting for ready %v", err)
+	}
+
+	if a.GetName() == "" {
+		return d.validateDeleteCollection(ctx, a)
+	}
+
+	return d.validateDelete(ctx, a)
+}
+
+func (d *ValidateShootStateDeletion) validateDeleteCollection(ctx context.Context, attrs admission.Attributes) error {
+	shootStateList, err := d.shootStateLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, shootState := range shootStateList {
+		if err := d.validateDelete(ctx, d.createAttributesWithName(shootState.Name, attrs)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *ValidateShootStateDeletion) validateDelete(ctx context.Context, attrs admission.Attributes) error {
+	_, err := d.coreClient.Core().Shoots(attrs.GetNamespace()).Get(attrs.GetName(), metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
```

Comment:
You should also handle other `err`s.

Reply:
I'll add a check for IsGone() all other errors should return forbidden IMHO.

Reply:
Shouldn't they return something like "InternalServerError". Why is `410 GONE` a special case?

Reply:
So you meant to add it so that the message is improved? 
410 GONE is handled the same way as 404 not Found -> the function returns nil

Reply:
updated in the latest commit to make errors like Internal Server Error transparent and directly return them to the user